### PR TITLE
[Breakponts] remove immutable.js wrappers

### LINKED ISSUE #7202: [breakpoints] remove immutable wrapper
When we started we used immutable to wrap our data so that we had guarantees that the data would only be updated in a reducer and that we could rely on the data being immutable, which made component updates cheap.

This summer we refactored sources away from immutable because immutable is hard to type check with flow, the api is harder to use, it is slower, and we have gotten good enough with `...` destructuring. We should now refactor breakpoints, which is the second biggest data source.

Steps:

1. change the reducer to be plain objects
2. run `yarn test` or `jest src/actions` and start fixing the unit tests
3. run `yarn flow` and start fixing the flow errors
4. pull out a good amount of hair 😢, ultimately achieve eternal wisdom 🙏 

This is a good amount of work, but will be a big help and teach you some interesting details about breakpoints in the debugger!

---

I started refactoring the reducer here

```diff
diff --git a/src/reducers/breakpoints.js b/src/reducers/breakpoints.js
index a86b5ac52..6cb9168e1 100644
--- a/src/reducers/breakpoints.js
+++ b/src/reducers/breakpoints.js
@@ -19,8 +19,8 @@ import type { XHRBreakpoint, Breakpoint, Location } from "../types";
 import type { Action, DonePromiseAction } from "../actions/types";
 import type { Record } from "../utils/makeRecord";
 
-export type BreakpointsMap = I.Map<string, Breakpoint>;
-export type XHRBreakpointsList = I.List<XHRBreakpoint>;
+export type BreakpointsMap = { [string]: Breakpoint };
+export type XHRBreakpointsList = XHRBreakpoint[];
 
 export type BreakpointsState = {
   breakpoints: BreakpointsMap,
@@ -29,14 +29,12 @@ export type BreakpointsState = {
 
 export function initialBreakpointsState(
   xhrBreakpoints?: any[] = []
-): Record<BreakpointsState> {
-  return makeRecord(
-    ({
-      breakpoints: I.Map(),
-      xhrBreakpoints: I.List(xhrBreakpoints),
-      breakpointsDisabled: false
-    }: BreakpointsState)
-  )();
+): BreakpointsState {
+  return {
+    breakpoints: {},
+    xhrBreakpoints: xhrBreakpoints,
+    breakpointsDisabled: false
+  };
 }
 
 function update(
@@ -148,11 +146,18 @@ function updateXHRBreakpoint(state, action) {
   return state.set("xhrBreakpoints", xhrBreakpoints.set(index, breakpoint));
 }
 
+function setBreakpoint(state, locationId, breakpoint) {
+  return {
+    ...state,
+    breakpoints: { ...state.breakpoints, [locationId]: breakpoint }
+  };
+}
+
 function addBreakpoint(state, action) {
   if (action.status === "start" && action.breakpoint) {
     const { breakpoint } = action;
     const locationId = makeLocationId(breakpoint.location);
-    return state.setIn(["breakpoints", locationId], breakpoint);
+    return setBreakpoint(state, locationId, breakpoint);
   }
 
   // when the action completes, we can commit the breakpoint
@@ -182,20 +187,20 @@ function syncBreakpoint(state, data) {
   }
 
   const locationId = makeLocationId(breakpoint.location);
-  return state.setIn(["breakpoints", locationId], breakpoint);
+  return setBreakpoint(state, locationId, breakpoint);
 }
 
 function updateBreakpoint(state, action) {
   const { breakpoint } = action;
   const locationId = makeLocationId(breakpoint.location);
-  return state.setIn(["breakpoints", locationId], breakpoint);
+  return setBreakpoint(state, locationId, breakpoint);
 }
 
 function updateAllBreakpoints(state, action) {
   const { breakpoints } = action;
   breakpoints.forEach(breakpoint => {
     const locationId = makeLocationId(breakpoint.location);
-    state = state.setIn(["breakpoints", locationId], breakpoint);
+    state = setBreakpoint(state, locationId, breakpoint);
   });
   return state;
}
```

### PR DESCRIPTION
**Pull request by kenjiO**
Fixes #7202

Updates breakpoint reducer to not use immutable.js wrappers.  Also updates anything that uses the reducer and selectors to work with the updated data structures.

### CODE REVIEW

**Comment by kenjiO** [comment]
@jasonLaster Can you review the first commit and let me know how this looks so far?  I noticed that most of the selectors use the breakpoints as an array rather than a map.  Would you prefer to store the breakpoints in the redux store as an array rather than a map to avoid converting it to an array?

File: src/actions/breakpoints/index.js [hunk 1]
```diff
@@ -179,9 +176,7 @@ export function toggleBreakpoints(
  */
 export function removeAllBreakpoints() {
   return async ({ dispatch, getState }: ThunkArgs) => {
-    const breakpointList = getBreakpoints(getState())
-      .valueSeq()
-      .toJS();
+    const breakpointList = (Object.values(getBreakpoints(getState())): any);
```
Comment (jasonLaster) [inline src/actions/breakpoints/index.js]: I think it would be nice to have a selector `getBreakpointList` that returns an array of breakpoints and can be flow typed

File: src/reducers/breakpoints.js [hunk 2]
```diff
@@ -29,20 +29,18 @@ export type BreakpointsState = {
...
-    ({
-      breakpoints: I.Map(),
-      xhrBreakpoints: I.List(xhrBreakpoints),
-      breakpointsDisabled: false
-    }: BreakpointsState)
-  )();
+): BreakpointsState {
+  return {
+    breakpoints: {},
+    xhrBreakpoints: I.List(xhrBreakpoints),
```
Comment (jasonLaster) [inline src/reducers/breakpoints.js]: we should probably leave this as immutable for now as that might be its own PR

**Comment by jasonLaster** [comment]
one lint error and needs a rebase. Lets get this in! :)

File: src/actions/breakpoints/index.js [hunk 3]
```diff
@@ -113,11 +113,11 @@ export function disableBreakpoint(location: Location) {
  */
 export function toggleAllBreakpoints(shouldDisableBreakpoints: boolean) {
   return async ({ dispatch, getState, client }: ThunkArgs) => {
-    const breakpoints = getBreakpoints(getState());
+    const breakpoints = getBreakpointsArray(getState());
```
Comment (jasonLaster) [inline src/actions/breakpoints/index.js]: this is a nit, but i think I prefer `getBreakpointsList` as it is more descriptive than implementation specific

File: src/actions/breakpoints/index.js [hunk 4]
```diff
@@ -157,15 +157,12 @@ export function toggleBreakpoints(
...
-    const promises = breakpoints
-      .valueSeq()
-      .toJS()
-      .map(
-        ([, breakpoint]) =>
-          shouldDisableBreakpoints
-            ? dispatch(disableBreakpoint(breakpoint.location))
-            : dispatch(enableBreakpoint(breakpoint.location))
-      );
+    const promises = (Object.values(breakpoints): any).map(
```
Comment (jasonLaster) [inline src/actions/breakpoints/index.js]: could we use `getBreakpointsArray`?

File: src/actions/breakpoints/index.js [hunk 5]
```diff
@@ -196,7 +191,7 @@ export function removeAllBreakpoints() {
  */
 export function removeBreakpoints(breakpoints: BreakpointsMap) {
   return async ({ dispatch }: ThunkArgs) => {
-    const breakpointList = breakpoints.valueSeq().toJS();
+    const breakpointList = (Object.values(breakpoints): any);
```
Comment (jasonLaster) [inline src/actions/breakpoints/index.js]: could we use `getBreakpointsArray`?

File: src/reducers/breakpoints.js [hunk 6]
```diff
@@ -118,12 +114,17 @@ function addXHRBreakpoint(state, action) {
   );
 
   if (existingBreakpointIndex === -1) {
-    return state.set("xhrBreakpoints", xhrBreakpoints.push(breakpoint));
+    // TODO make this immutable?
+    return {
```
Comment (jasonLaster) [inline src/reducers/breakpoints.js]: it is okay to have mutable state as long as you copy it and ensure that it does not change. We ensure this with source records via flow 

e.g. `type Source = {| +a: 2 |}; let s: Source = {a: 2}`. Now `a` is read only :)

File: src/reducers/breakpoints.js [hunk 7]
```diff
@@ -214,60 +246,73 @@ function remapBreakpoints(state, action) {
...
 }
 
 // Selectors
 // TODO: these functions should be moved out of the reducer
 
-type OuterState = { breakpoints: Record<BreakpointsState> };
+type OuterState = { breakpoints: BreakpointsState };
 
-export function getBreakpoints(state: OuterState) {
+// TODO make this immutable?
```
Comment (jasonLaster) [inline src/reducers/breakpoints.js]: i don't think it needs to be immutable as long as we prevent writes via flow

File: src/reducers/breakpoints.js [hunk 8]
```diff
@@ -214,60 +246,73 @@ function remapBreakpoints(state, action) {
...
+export function getBreakpointsMap(state: OuterState): BreakpointsMap {
   return state.breakpoints.breakpoints;
 }
 
+export function getBreakpointsArray(state: OuterState): Breakpoint[] {
```
Comment (jasonLaster) [inline src/reducers/breakpoints.js]: I think this would read a bit nicer:

```js
 export function getBreakpointsList(state: OuterState): Breakpoint[] { 
```

File: src/reducers/breakpoints.js [hunk 9]
```diff
@@ -214,60 +246,73 @@ function remapBreakpoints(state, action) {
...
   }
 
   const isGeneratedSource = isGeneratedId(sourceId);
-  const breakpoints = getBreakpoints(state);
+  const breakpoints = getBreakpointsMap(state);
 
-  return breakpoints.filter(bp => {
+  const breakpointsForSource = {};
+  const keys = Object.keys(breakpoints);
+  keys.forEach(key => {
```
Comment (jasonLaster) [inline src/reducers/breakpoints.js]: i think it might be nicer to do `for (const key in breakpoints)`

File: src/reducers/tests/breakpoints.spec.js [hunk 10]
```diff
@@ -11,20 +11,66 @@ import {
...
-  return state.set("breakpoints", mappedData);
+  state.breakpoints = data;
+  return state;
 }
 
 describe("Breakpoints Selectors", () => {
   it("it gets a breakpoint for an original source", () => {
     const sourceId = "server1.conn1.child1/source1/originalSource";
-    const matchingBreakpoints = { id1: { location: { sourceId } } };
+    const matchingBreakpoints = {
```
Comment (jasonLaster) [inline src/reducers/tests/breakpoints.spec.js]: why is this so different?

File: src/selectors/breakpointSources.js [hunk 11]
```diff
@@ -67,10 +66,7 @@ function findBreakpointSources(
   breakpoints: BreakpointsMap
 ): Source[] {
   const sourceIds: string[] = uniq(
-    breakpoints
-      .valueSeq()
-      .map(bp => getSelectedLocation(bp, selectedSource).sourceId)
-      .toJS()
+    (Object.values(breakpoints): any).map(bp => bp.location.sourceId)
```
Comment (jasonLaster) [inline src/selectors/breakpointSources.js]: can we use the new selector?

File: src/test/mochitest/browser_dbg-editor-gutter.js [hunk 12]
```diff
@@ -38,20 +38,20 @@ add_task(async function() {
...
   await waitForDispatch(dbg, "ADD_BREAKPOINT");
-  is(getBreakpoints(getState()).size, 1, "One breakpoint exists");
+  is(getBreakpointsArray(getState()).length, 1, "One breakpoint exists");
   assertEditorBreakpoint(dbg, 4, true);
 
   // Make sure clicking at the same place removes the icon.
   clickGutter(dbg, 4);
   await waitForDispatch(dbg, "REMOVE_BREAKPOINT");
-  is(getBreakpoints(getState()).size, 0, "No breakpoints exist");
+  is(getBreakpointsArray(getState()).length, 0, "No breakpoints exist");
```
Comment (jasonLaster) [inline src/test/mochitest/browser_dbg-editor-gutter.js]: I see this cropping up a lot, should we have a selector `getBreakpointCount`?

File: src/actions/breakpoints/index.js (hunk 4 above)
Comment (kenjiO) [reply src/actions/breakpoints/index.js]: It looks like the code that calls this function is already passing in an array instead of a map so I can just change the type and drop the Object.values() call

File: src/reducers/tests/breakpoints.spec.js (hunk 10 above)
Comment (kenjiO) [reply src/reducers/tests/breakpoints.spec.js]: flow complained about all the missing properties so I added them

**Comment by jasonLaster** [comment]
Looks great. If it's green, I'll do some QA testing and fully expect to press the green button tomorrow.

**Comment by jasonLaster** [comment]
It would be nice to update the Flow types for Breakpoints now as well and make sure we *actually* are immutable. I'd be nervous landing this without some guarantees that we have flow coverage to make sure we don't accidentally mutate outside of the reducers.

```diff
diff --git a/src/types.js b/src/types.js
index 961b136fa..aebdc18a1 100644
--- a/src/types.js
+++ b/src/types.js
@@ -93,31 +93,31 @@ export type ASTLocation = {|
  * @memberof types
  * @static
  */
-export type Breakpoint = {
-  id: BreakpointId,
-  location: Location,
-  astLocation: ?ASTLocation,
-  generatedLocation: Location,
-  loading: boolean,
-  disabled: boolean,
-  hidden: boolean,
-  text: string,
-  originalText: string,
-  condition: ?string
-};
+export type Breakpoint = {|
+  +id: BreakpointId,
+  +location: Location,
+  +astLocation: ?ASTLocation,
+  +generatedLocation: Location,
+  +loading: boolean,
+  +disabled: boolean,
+  +hidden: boolean,
+  +text: string,
+  +originalText: string,
+  +condition: ?string
+|};

 /**
  * XHR Breakpoint
  * @memberof types
  * @static
  */
-export type XHRBreakpoint = {
-  path: string,
-  method: "GET" | "POST" | "DELETE" | "ANY",
-  loading: boolean,
-  disabled: boolean,
-  text: string
-};
+export type XHRBreakpoint = {|
+  +path: string,
+  +method: "GET" | "POST" | "DELETE" | "ANY",
+  +loading: boolean,
+  +disabled: boolean,
+  +text: string
+|};
```

File: src/selectors/visibleBreakpoints.js [hunk 13]
```diff
@@ -46,8 +45,8 @@ function isVisible(breakpoint, selectedSource) {
   */
 export const getVisibleBreakpoints = createSelector(
```
Comment (kenjiO) [inline src/selectors/visibleBreakpoints.js]: Do you know how to add a flow type to getVisibleBreakpoints?  I think the issue with the "disable other breakpoints" has to do with a Breakpoint being compared to a FormattedBreakpoint from this selector that gets a false negative.

Comment (jasonLaster) [reply src/selectors/visibleBreakpoints.js]: hmm, i forgot. what do we we do with other memozed selectors?

**Comment by jasonLaster** [comment]
@kenjiO i just updated this PR w/ some small tweaks and squashed your commits down.

File: src/reducers/tests/breakpoints.spec.js [hunk 14]
```diff
@@ -11,20 +11,24 @@ import {
...
+      id1: createBreakpoint({ line: 1, sourceId: sourceId })
+    };
+
     const otherBreakpoints = {
-      id2: { location: { sourceId: "not-this-source" } }
+      id2: createBreakpoint({ line: 1, sourceId: "not-this-source" })
```
Comment (jasonLaster) [inline src/reducers/tests/breakpoints.spec.js]: this is a fun trick to keep the fixture data lean. We do the same thing for sources in some other tests

Comment (kenjiO) [reply src/reducers/tests/breakpoints.spec.js]: Cool, I will update the tests with that

**Comment by kenjiO** [comment]
RE: 8f47451, can we add these '+' to the types?  I tried adding the '|' to the types but that got a lot of flow errors.  Maybe save that for another PR.  I still haven't fixed the issue with 'disable other breakpoints' yet so please don't merge it unless you fixed it.

**Comment by jasonLaster** [comment]
> disable other breakpoints

thanks for the heads up. I definitely want to land this oon.

> other `+` 

thanks!

**Comment by jasonLaster** [comment]
> disable other breakpoints

so i think this is a big enough change, where we likely will have some small bugs. Lets land this and then fix the disable other breakpoints bug. Maybe add a unit test for it as well.

**Comment by kenjiO** [comment]
@jasonLaster I think this should fix it.  I'll work on some tests for it.  Should I do jest tests, mochi tests or both?

```diff
diff --git a/src/components/SecondaryPanes/Breakpoints/BreakpointsContextMenu.js b/src/components/SecondaryPanes/Breakpoints/BreakpointsContextMenu.js
index 510ef51a..09b7bd42 100644
--- a/src/components/SecondaryPanes/Breakpoints/BreakpointsContextMenu.js
+++ b/src/components/SecondaryPanes/Breakpoints/BreakpointsContextMenu.js
@@ -4,6 +4,20 @@
 
 import { buildMenu, showMenu } from "devtools-contextmenu";
 
+/**
+ * Check if a Breakpoint type object and a FormattedBreakpoint type object
+ * are the same breakpoint
+ */
+function checkIfSameBreakpoint(breakpoint, formattedBreakpoint) {
+  const location1 = breakpoint.location;
+  const location2 = formattedBreakpoint.selectedLocation;
+  return (
+    location1.sourceId === location2.sourceId &&
+    location1.line === location2.line &&
+    location1.column === location2.column
+  );
+}
+
 export default function showContextMenu(props) {
   const {
     removeBreakpoint,
@@ -74,14 +88,16 @@ export default function showContextMenu(props) {
     "breakpointMenuItem.addCondition2.accesskey"
   );
 
-  const otherBreakpoints = breakpoints.filter(b => b !== breakpoint);
+  const otherBreakpoints = breakpoints.filter(
+    b => !checkIfSameBreakpoint(b, breakpoint)
+  );
   const enabledBreakpoints = breakpoints.filter(b => !b.disabled);
   const disabledBreakpoints = breakpoints.filter(b => b.disabled);
   const otherEnabledBreakpoints = breakpoints.filter(
-    b => !b.disabled && b !== breakpoint
+    b => !b.disabled && !checkIfSameBreakpoint(b, breakpoint)
   );
   const otherDisabledBreakpoints = breakpoints.filter(
-    b => b.disabled && b !== breakpoint
+    b => b.disabled && !checkIfSameBreakpoint(b, breakpoint)
   );
 
   const deleteSelfItem = {
```

**Comment by jasonLaster** [comment]
lets move this patch to a new WIP pr.

My first thoughts are you could use `bp.id` to compare values. Also, why does `breakpoint != b` fail? we should be returning the same breakpoint? 

Lines like [this](https://github.com/devtools-html/debugger.html/blob/master/src/reducers/breakpoints.js#L309-L310) where we are creating new objects when nothing changed are probably a mistake. Actually, it could be a big perf regression if this means components will start re-rendering because they think its a new object :/

**Comment by jasonLaster** [comment]
it looks like the breakpoint patch has a performance regression... do you have some time to dig in and investigate?

**Comment by kenjiO** [comment]
@jasonLaster Yes, I will look into it.  What tool do you reccommend to measure peformance and is there a place I kind find more details about the performance regression?

**Comment by jasonLaster** [comment]
>  What tool do you reccommend to measure peformance

http://perf-html.io/ and chrome devtools performance

> is there a place I kind find more details about the performance regression?

not really...
